sdks/java/core/src/main/java/org/apache/beam/sdk/options/ProxyInvocationHandler.java

```diff
@@ -265,88 +264,96 @@ public int hashCode() {
-        continue;
-      }
-
-      HashSet<PipelineOptionSpec> specs = new HashSet<>(optionsMap.get(jsonOption.getKey()));
-      if (specs.isEmpty()) {
-        builder.add(DisplayData.item(jsonOption.getKey(), jsonOption.getValue().toString())
-          .withNamespace(UnknownPipelineOptions.class));
-      } else {
-        for (PipelineOptionSpec spec : specs) {
-          if (!spec.shouldSerialize()) {
+        for (PipelineOptionSpec optionSpec : specs) {
+          if (!optionSpec.shouldSerialize()) {
+            // Options that are excluded for serialization (i.e. those with @JsonIgnore) are also
+            // excluded from display data. These options are generally not useful for display.
             continue;
           }
 
-          Object value = getValueFromJson(jsonOption.getKey(), spec.getGetterMethod());
-          value = value == null ? "" : value;
-          DisplayData.Type type = DisplayData.inferType(value);
+          Class<?> pipelineInterface = optionSpec.getDefiningInterface();
           if (type != null) {
-            builder.add(DisplayData.item(jsonOption.getKey(), type, value)
-                .withNamespace(spec.getDefiningInterface()));
+            builder.add(DisplayData.item(option.getKey(), type, value)
+                .withNamespace(pipelineInterface));
           } else {
-            builder.add(DisplayData.item(jsonOption.getKey(), displayDataString(value))
-                .withNamespace(spec.getDefiningInterface()));
+            builder.add(DisplayData.item(option.getKey(), displayDataString(value))
+                .withNamespace(pipelineInterface));
           }
         }
       }
-    }
-  }
 
-  /**
-   * {@link Object#toString()} wrapper to extract display data values for various types.
-   */
-  private String displayDataString(Object value) {
-    checkNotNull(value, "value cannot be null");
-    if (!value.getClass().isArray()) {
-      return value.toString();
-    }
-    if (!value.getClass().getComponentType().isPrimitive()) {
-      return Arrays.deepToString((Object[]) value);
+      for (Map.Entry<String, JsonNode> jsonOption : jsonOptions.entrySet()) {
+        if (options.containsKey(jsonOption.getKey())) {
+          // Option overwritten since deserialization; don't re-write
+          continue;
+        }
+
+        HashSet<PipelineOptionSpec> specs = new HashSet<>(optionsMap.get(jsonOption.getKey()));
+        if (specs.isEmpty()) {
+          builder.add(DisplayData.item(jsonOption.getKey(), jsonOption.getValue().toString())
+              .withNamespace(UnknownPipelineOptions.class));
+        } else {
+          for (PipelineOptionSpec spec : specs) {
+            if (!spec.shouldSerialize()) {
+              continue;
+            }
+
+            Object value = getValueFromJson(jsonOption.getKey(), spec.getGetterMethod());
+            value = value == null ? "" : value;
+            DisplayData.Type type = DisplayData.inferType(value);
```

Comment:
Seems to be a little duplication between here and above (eg., up around 299). Anyway to extract some of this to a helper method?